h2o-algos/src/main/java/hex/gam/GamSplines/ThinPlateRegressionUtils.java

```diff
@@ -0,0 +1,347 @@
+import static water.util.ArrayUtils.maxValue;
+
+// This class contains functions that perform different functions for generating the thin plate regression splines
+// and the polynomial basis functions
+public class ThinPlateRegressionUtils {
+  /***
+   * For thin plate regression, given d (number of predictors for a smooth), it will return m where (m-1) is the
+   * maximum polynomial degree in the polynomial basis functions.  The formula used is m = ceiling of (d+1)/2+1
+   *
+   * @param d : integer denoting number of predictors for thin plate regression smooth.
+   * @return m : integer denoting the maximum polynomial degree + 1 for polynomial basis function.
+   */
+  public static int calculatem(int d) {
+    return ((int) Math.floor((d+1.0)*0.5))+1;
+  }
+
+  public static int calculateM(int d, int m) {
+    int topComb = d+m-1;
+    return hex.genmodel.utils.MathUtils.combinatorial(topComb, d);
+  }
+
+  /***
+   * This method, given number of predictors in the smooth d, number of polynomials in the polynomial basis m, will
+   * generate a list of integer array specifying for each predictors the degree that predictor will have.  For instance,
+   * if for a predictor, the degree is 0, a constant 1 is used.  If for a particular predictor, the degree is 2, 
+   * predictor*predictor is used.
+   * 
+   * @param d
+   * @param m
+   * @return
+   */
+  public static List<Integer[]> findPolyBasis(int d, int m) {
+    int polyOrder = m-1;
+    int[] possibleDegree = new int[polyOrder];
+    for (int index = 1; index < m; index++)             // generate all polynomial order combinations
+      possibleDegree[index-1] = index;
+    Integer[] basisPolyOrder = new Integer[d];          // store one combination
+    List<Integer[]> totPolyBasis = new ArrayList<>();   // store all combination 
+    for (int degree : possibleDegree) {
+      ArrayList<int[]> oneCombo = new ArrayList<>();
+      findOnePerm(degree, possibleDegree, 0, oneCombo, null);
+      mergeCombos(oneCombo, basisPolyOrder, possibleDegree, totPolyBasis);// merge all combos found for all possibleDegree
+    }
+    return findAllPolybasis(totPolyBasis);
+  }
+
+  /***
+   * For each list in onePolyBasis, we still need to find all the permutations for that list.  In addition, we need to
+   * add the combination for the 0th order as well.  For instance, if the list contains {0,0,1}, we need to add to that
+   * the lists {0,1,0} and {1,0,0} as well.
+   * @param onePolyBasis
+   * @return
+   */
+  public static List<Integer[]> findAllPolybasis(List<Integer[]> onePolyBasis) {
+    int listSize = onePolyBasis.size();
+    List<Integer[]> allPermutes = new ArrayList<>();
+    for (int index = 0; index < listSize; index++) {
+      Integer[] oneBasis = onePolyBasis.get(index);
+      int[] freqTable = generateOrderFreq(oneBasis);  // find polynomial basis order and count
+      ArrayList<ArrayList<Integer>> basisPermuations = new ArrayList<>();
```

Comment:
`List` interface should be used for variable type declaration, already used on some places but in some not

Reply:
Done.  Thanks.